From c0c89705f976053cc9880cbea8ef4f4248cfb95f Mon Sep 17 00:00:00 2001
From: Jakompiled <itsrufer@gmail.com>
Date: Wed, 7 Oct 2026 15:01:36 +0000
Subject: [PATCH 01/11] IOP kernel: keep the real signature of thread entry
 points

IOP thread entries are u32 (*)() but were stored as void (*)() and called
through that type. This works on x86 and ARM, but WebAssembly checks the
signature of every indirect call and traps.
---
 game/sce/iop.cpp           | 2 +-
 game/system/IOP_Kernel.cpp | 6 +++---
 game/system/IOP_Kernel.h   | 6 +++---
 3 files changed, 7 insertions(+), 7 deletions(-)

diff --git a/game/sce/iop.cpp b/game/sce/iop.cpp
index 6756fa15..1b18f812 100644
--- a/game/sce/iop.cpp
+++ b/game/sce/iop.cpp
@@ -96,7 +96,7 @@ void* AllocScratchPad(int mode) {
  * Create a new thread
  */
 s32 CreateThread(ThreadParam* param) {
-  return iop->kernel.CreateThread(param->name, (void (*)())param->entry, param->initPriority);
+  return iop->kernel.CreateThread(param->name, param->entry, param->initPriority);
 }
 
 /*!
diff --git a/game/system/IOP_Kernel.cpp b/game/system/IOP_Kernel.cpp
index 04e6f953..36c037b5 100644
--- a/game/system/IOP_Kernel.cpp
+++ b/game/system/IOP_Kernel.cpp
@@ -14,10 +14,10 @@ using namespace std::chrono;
 /*
 ** wrap thread entry points to ensure they don't return into libco
 */
-static void (*thread_entry)() = nullptr;
+static u32 (*thread_entry)() = nullptr;
 static cothread_t wrap_return;
 void IopThread::functionWrapper() {
-  void (*f)() = thread_entry;
+  u32 (*f)() = thread_entry;
   co_switch(wrap_return);
   if (f != nullptr) {
     f();
@@ -52,7 +52,7 @@ u32 IOP_Kernel::GetSystemTimeLow() {
 /*!
  * Create a new thread.  Will not run the thread.
  */
-s32 IOP_Kernel::CreateThread(std::string name, void (*func)(), u32 priority) {
+s32 IOP_Kernel::CreateThread(std::string name, u32 (*func)(), u32 priority) {
   u32 ID = (u32)_nextThID++;
   ASSERT(ID == threads.size());
 
diff --git a/game/system/IOP_Kernel.h b/game/system/IOP_Kernel.h
index cf3e9fde..ce17c375 100644
--- a/game/system/IOP_Kernel.h
+++ b/game/system/IOP_Kernel.h
@@ -55,7 +55,7 @@ struct IopThread {
 
   enum class Wait { None, Sleep, Semaphore, Delay, Messagebox, EventFlag };
 
-  IopThread(std::string n, void (*f)(), s32 ID, u32 pri)
+  IopThread(std::string n, u32 (*f)(), s32 ID, u32 pri)
       : name(std::move(n)), function(f), priority(pri), thID(ID) {
     thread = co_create(0x300000, functionWrapper);
   }
@@ -64,7 +64,7 @@ struct IopThread {
 
   static void functionWrapper();
   std::string name;
-  void (*function)();
+  u32 (*function)();  // real signature of IOP thread entries (wasm checks call types)
   cothread_t thread;
   State state = State::Dormant;
   Wait waitType = Wait::None;
@@ -112,7 +112,7 @@ struct Messagebox {
 class IOP_Kernel {
  public:
   IOP_Kernel();
-  s32 CreateThread(std::string n, void (*f)(), u32 priority);
+  s32 CreateThread(std::string n, u32 (*f)(), u32 priority);
   s32 ExitThread();
   void StartThread(s32 id);
   void DelayThread(u32 usec);
-- 
2.43.0

